File: apps/backend/src/models/Credential.ts
import { Schema, model, Document } from 'mongoose';
import type { CredentialClaim } from '@onshift/shared-types';

export interface CredentialDocument extends Document {
  type: string;
  credentialType?: string;
  issuer: string;
  issuerPublicKey?: string;
  publicKeyHex?: string;
  workerId: string;
  verificationId?: string;
  issuedAt: string;
  validUntil: string;
  claims: CredentialClaim;
  signature: string;
}

const CredentialSchema = new Schema<CredentialDocument>(
  {
    type: {
      type: String,
      required: true,
      default: 'OnShiftIncomeCredential',
    },
    credentialType: {
      type: String,
      required: false,
    },
    issuer: {
      type: String,
      required: true,
      default: 'OnShift Proof Authority',
    },
    issuerPublicKey: {
      type: String,
      required: false,
    },
    publicKeyHex: {
      type: String,
      required: true,
    },
    workerId: {
      type: String,
      required: true,
      index: true,
    },
    verificationId: {
      type: String,
      required: false,
      index: true,
    },
    issuedAt: {
      type: String,
      required: true,
    },
    validUntil: {
      type: String,
      required: true,
    },
    claims: {
      verifiedIncome: { type: Number, required: true },
      period: { type: String, required: true },
      verificationLevel: {
        type: String,
        required: true,
        enum: ['DECLARED', 'OBSERVED', 'CORROBORATED', 'FINANCIALLY_CORROBORATED'],
      },
      platformBreakdown: { type: Schema.Types.Mixed, required: false },
      identityVerified: { type: Boolean, required: false },
    },
    signature: {
      type: String,
      required: true,
    },
  },
  {
    timestamps: false,
    versionKey: false,
  }
);

// Enforce idempotency: A worker can only have one credential issued per verificationId
CredentialSchema.index({ workerId: 1, verificationId: 1 }, { unique: true, sparse: true });

export const Credential = model<CredentialDocument>('Credential', CredentialSchema);

File: apps/android/app/src/main/java/com/onshift/app/data/model/Credential.kt
package com.onshift.app.data.model

data class Credential(
    val type: String = "OnShiftIncomeCredential",
    val workerId: String,
    val issuer: String = "OnShift Proof Authority",
    val issuedAt: String = "",
    val validUntil: String = "",
    val period: String = "01 Aug to 07 Aug 2026",
    val verifiedIncome: Double? = 30100.0,
    val verificationLevel: VerificationLevel? = VerificationLevel.FINANCIALLY_CORROBORATED,
    val signaturePreview: String? = null,
    val signature: String? = null,
    val publicKeyHex: String? = null,
    val verificationId: String? = null,
    val includedClaims: List<String> = emptyList(),
    val claims: Map<String, Any?>? = null
)
